### DISCUSSION TITLE
Failed LC Geometry solution

### DISCUSSION THREAD

**Discussion by thillias** [phoebe-project/phoebe2#803] [Questions]
Hi

For a university project, I have received light curve and radial velocity data of an unspecified binary system. I should use PHOEBE to constrain different parameters given this data. As you might have guessed, I have not had much success at doing so. 

At the moment, I'm using the LC geometry solver for my light curve data. I have also tried both EBAI solvers. I managed to flip constraints and run the solver but the solution it offers me seems very inaccurate (see below):

![LCGEOM](https://github.com/phoebe-project/phoebe2/assets/100118070/d919e37e-f7e7-4aca-8210-82bc89ef63d9)

This fit is obtained using the polyfit method since the two-gaussian offered nan values. Note that the flux drops are very narrow in shape and as such the knn EBAI method is probably not the best choice either way.

For the RV data, I'm using the RV geometry solver. In this case, I'm having less success in running the solver. It yields the "ValueError: Residuals are not finite in the initial point." error.

I'm probably missing obvious things? Please let me know if seeing my code helps, I'll be happy to offer it.

Thanks in advance

**Comment by kecnry** [comment]
It doesn't surprise me that the estimators would struggle here since there isn't great signal to noise.  You can play with some of the input options (binning and 2-gaussian vs polyfit), but ultimately if the proposed parameters or the exposed model fit don't look correct, then I would suggest just not adopting the proposed changes and instead estimating starting positions for your parameters manually. 

> For the RV data, I'm using the RV geometry solver. In this case, I'm having less success in running the solver. It yields the "ValueError: Residuals are not finite in the initial point." error.

It's hard to tell without seeing the data, but my guess is that one of the bins is completely empty of data.  If that's the case, you can try decreasing the number of bins or disable binning entirely and hopefully that should get around the error.  We are planning to eventually improve handling of these cases (see #749), but it is a little tricky to get right for all scenarios.  If you're still getting an error when disabling binning, check your input RVs and uncertainties to see if there are any infinities or nans, or if you're willing to post the data/script/bundle someone can try to take a closer look to see what is happening.

Hope this helps!

**Reply by thillias** [reply]
Hi Kyle 

Thanks for the quick response. I have shifted my attention away from trying the estimators to work; I think you're right on the SNR front. I'm now trying to run the Nelder-Mead optimizer on my LC/RV data, again without much success. I do have to note that my system does not have the ellc module installed yet, since I'm not qualified to install this myself (I should wait for someone to do it on Monday). However, as far as my understanding goes, the ellc module is mainly used in plotting the optimized curve using a forward model? As such, I should be able to do optimization without ellc for now?

Either way, after a significant amount of iterations (2000), the optimized values for my parameters simply remain equal to my initial given values. 

Concerning the RV data, I have turned off binning completely and there are no nan/inf values in my data.

For those who feel like it, the data I have gotten assigned is attached. If in need of the precise code I am using, please let me know.

I'm attempting to obtain optimized values for the following parameters : vgamma, period, eccentricity, Teffratio, inclination, sma, q and per0
and have obtained following manual estimates for them:
-20 km/s, 23.94 days, 0.3, 1.23, 85deg, 53.2 Rsol, 0.469 and 108deg

Note that the mass ratio q is smaller than 1 here. This is the way the data was given to me. In the data, the primary is denoted as the less massive star of the two.

Forgive any obvious mistakes, I'm fairly new to PHOEBE.

Thanks

**Reply by kecnry** [reply]
> I do have to note that my system does not have the ellc module installed yet, since I'm not qualified to install this myself (I should wait for someone to do it on Monday). However, as far as my understanding goes, the ellc module is mainly used in plotting the optimized curve using a forward model? As such, I should be able to do optimization without ellc for now?

ellc is a different eclipsing binary modeling code, so when used in phoebe, it just swaps out phoebe's light curve and RV synthesis with ellc's (which makes [different assumptions](http://phoebe-project.org/docs/2.4/examples/compute_comparison_table)).

> after a significant amount of iterations (2000), the optimized values for my parameters simply remain equal to my initial given values.

This doesn't sound right.  Are you using any priors?  What is the cost function at the starting point?  (Your data didn't seem to attach successfully, it can be a little finicky to attach files in GitHub)

**Reply by thillias** [reply]
Second attempt at attaching the RV and LC data. Hopefully this time's the charm.
[LC_data_unfolded.txt](https://github.com/phoebe-project/phoebe2/files/13599838/LC_data_unfolded.txt)
[RV_data.txt](https://github.com/phoebe-project/phoebe2/files/13599861/RV_data.txt)

We are using priors that are distributed by a Gaussian. The cost function is just the default one, calculate_lnlikelihood if I'm not mistaken.

Attaching some code images below : 

![image](https://github.com/phoebe-project/phoebe2/assets/100118070/e295997e-ada6-48bc-968b-efb5021458f0)
![image](https://github.com/phoebe-project/phoebe2/assets/100118070/f3d7a2e9-b79e-4d39-a11f-72b8f9849cdd)

Thanks

**Reply by kecnry** [reply]
By adding the priors, you're telling the optimizer to favor those initial values.  I would suggest not using priors unless the optimizer is wandering into parameter-space that is unphysical, and if you set `expose_lnprobabilities=True`, you'll be able to see the cost function value at the beginning and end of the optimization run (or you can call calculate_chi2 after run_compute to see the current chi2 - if for some reason it is not finite, then that could explain the optimizer getting "stuck").

**Reply by thillias** [reply]
I have walked away from using the priors and indeed see infinite behaviour in my initial lnprobability. Any common reasons for this you know of from experience?

**Reply by kecnry** [reply]
It looks like one of your RV points has a zero uncertainty which would cause an infinite chi2. Hopefully that's the only cause, but if it's still returning infinities, I can try to take a closer look.

**Reply by thillias** [reply]
My bad for not specifying. The given RV data in the assignment does not come with errors. The 0/1 columns you see are the relative weights, as indicated in the readme file supplied with the data:
[data_H_new (1).zip](https://github.com/phoebe-project/phoebe2/files/13604682/data_H_new.1.zip)

**Reply by thillias** [reply]
Was just told to assign artificial errors of \pm 5 km/s. Adding the RV dataset with these values doesn't help, the initial lnprobability remains minus infinity. Not sure why though.

**Reply by kecnry** [reply]
Can you run a forward model (b.run_compute) at that same initial position and double check that the residuals are all finite (via [b.calculate_residuals](http://phoebe-project.org/docs/latest/api/phoebe.parameters.ParameterSet.calculate_residuals))?

**Reply by thillias** [reply]
Hi. Calculate_residuals requires a model argument. I'm however not sure what model Nelder-Mead uses to fit for example RV data?

**Reply by thillias** [reply]
When using nm_model, I do obtain finite residuals. The initial residuals in the nm_solver however remain infinite? 
![afbeelding](https://github.com/phoebe-project/phoebe2/assets/100118070/13eeef10-ec0d-41a7-8c3d-e43335df4f3c)

**Reply by thillias** [reply]
By "initial residuals" I am referring to the ln probabilities

**Reply by kecnry** [reply]
If you can save (and attach) the bundle directly before you would call the run_solver command that fails, we can take a look and try to sort out what is happening.

**Reply by thillias** [reply]
[PhoebeBundle.txt](https://github.com/phoebe-project/phoebe2/files/13676314/PhoebeBundle.txt)

Please find the bundle attached. Do note that I have tried the differential evolution optimizer, which does not struggle with infinite residuals, and have obtained an optimized set of parameters. The results are however fairly far from the data.

**Reply by kecnry** [reply]
That bundle does not contain the Nelder-Mead optimizer options, can you attach the bundle directly before calling `b.run_solver(solver='nm_solver')` (or whatever you had it labeled) that was resulting in non-finite probabilities?

**Reply by thillias** [reply]
[PhoebeBundlev2.txt](https://github.com/phoebe-project/phoebe2/files/13676634/PhoebeBundlev2.txt)

**Reply by kecnry** [reply]
Running `b.run_solver(solver='nm_solver')` _does_ work on my end, giving an initial lnprobability of -2313 and a final lnprobability of -850.

nm_solver references using the 'fastcompute' set of compute options - does `b.run_compute(compute='fastcompute')` work for you?  Do you have ellc installed?

**Reply by thillias** [reply]
Hi Kyle. I have indeed managed to also obtain finite lnprobabilities by combining information of some other discussion posts, like adopting a phoenix atmosphere everywhere. (Phoenix since I read that it is not recommended to use blackbody)

I do not have ellc installed since I'm not permitted to do so, but I can give asking another go. Is it necessary for optimization?

To summarize : I'm now at a point where I'm able to perform optimization on RV and LC data using the Nelder-Mead method. The optimization procedure still however always reaches max iterations and the LC fit seems quite off. The largest I went was 150 iterations and it took about 2 hours. I'm fitting parameters q, vgamma, t0supconj, ecc, per0, asini and teffratio based on the LC and RV data (see below). Any ideas on the reason for the LC fit failure? I would guess this is due to a bad value for the eccentricity and the teffratio?

![RV](https://github.com/phoebe-project/phoebe2/assets/100118070/76f25f99-2593-4ae6-a501-4d9490f1752c)
![LC](https://github.com/phoebe-project/phoebe2/assets/100118070/c00a4c92-94f1-4acd-b03d-bf7ccb6fa4d6)

**Reply by kecnry** [reply]
The bundle you sent me had the compute options selected to run ellc, you of course don't need to use that however.  Your initial guess will need to be much closer (eclipse positions at the correct phase, etc) in order for the optimizer to have enough information to converge to a solution.  Estimators can help, but ultimately will either need to do a large search (with something like nested sampling which is extremely expensive) or manually modify the parameter values until the synthetic light curve roughly resembles the observations.

**Reply by thillias** [reply]
Good point indeed, have been working on improving the parameter guesses. Any idea on how to determine the eclipse edges, it seems to be an output of the LC geometry estimator? In any case I can't seem to find it as a parameter in the binary bundle.

**Reply by kecnry** [reply]
The LC geometry estimator does expose the values that it finds, but in your case that seemed to not do well because of the low signal to noise.  I think you'll need to just do it manually by eye and once you get the eclipses in the approximate location and depth, then optimizers will be much more effective.
